File: ruby/lib/woothee.rb
# -*- coding: utf-8 -*-

require 'woothee/dataset'
require 'woothee/browser'
require 'woothee/os'
require 'woothee/mobilephone'
require 'woothee/crawler'
require 'woothee/appliance'
require 'woothee/misc'

module Woothee
  def self.parse(useragent)
    fill_result(exec_parse(useragent))
 end

  def self.is_crawler(useragent)
    useragent.length > 0 and useragent != '-' and try_crawler(useragent, {})
  end

  def self.exec_parse(useragent)
    result = {}

    return result if useragent.length < 1 or useragent == '-'

    if try_crawler(useragent, result)
      return result
    end

    if try_browser(useragent, result)
      if try_os(useragent, result)
        return result
      else
        return result
      end
    end

    if try_mobilephone(useragent, result)
      return result
    end

    if try_appliance(useragent, result)
      return result
    end

    if try_misc(useragent, result)
      return result
    end

    # browser unknown. check os only
    if try_os(useragent, result)
      return result
    end

    if try_rare_cases(useragent, result)
      return result
    end

    result
  end

  def self.try_crawler(useragent, result)
    return true if Woothee::Crawler.challenge_google(useragent, result)

    return true if Woothee::Crawler.challenge_crawlers(useragent, result)

    false
  end

  def self.try_browser(useragent, result)
    return true if Woothee::Browser.challenge_msie(useragent, result)

    return true if Woothee::Browser.challenge_safari_chrome(useragent, result)

    return true if Woothee::Browser.challenge_firefox(useragent, result)

    return true if Woothee::Browser.challenge_opera(useragent, result)

    false
  end

  def self.try_os(useragent, result)
    return true if Woothee::OS.challenge_windows(useragent, result)

    # OSX PC and iOS devices (strict check)
    return true if Woothee::OS.challenge_osx(useragent, result)

    # Linux PC and Android
    return true if Woothee::OS.challenge_linux(useragent, result)

    # all useragents matches /(iPhone|iPad|iPod|Android|BlackBerry)/
    return true if Woothee::OS.challenge_smartphone(useragent, result)

    # mobile phones like KDDI-.*
    return true if Woothee::OS.challenge_mobilephone(useragent, result)

    # Nintendo DSi/Wii with Opera
    return true if Woothee::OS.challenge_appliance(useragent, result)

    # Win98, BSD, classic MacOS, ...
    return true if Woothee::OS.challenge_misc(useragent, result)

    false
  end

  def self.try_mobilephone(useragent, result)
    return true if Woothee::MobilePhone.challenge_docomo(useragent, result)

    return true if Woothee::MobilePhone.challenge_au(useragent, result)

    return true if Woothee::MobilePhone.challenge_softbank(useragent, result)

    return true if Woothee::MobilePhone.challenge_willcom(useragent, result)

    return true if Woothee::MobilePhone.challenge_misc(useragent, result)

    false
  end

  def self.try_appliance(useragent, result)
    return true if Woothee::Appliance.challenge_playstation(useragent, result)

    return true if Woothee::Appliance.challenge_nintendo(useragent, result)

    return true if Woothee::Appliance.challenge_digitaltv(useragent, result)

    false
  end

  def self.try_misc(useragent, result)
    return true if Woothee::Misc.challenge_desktoptools(useragent, result)

    false
  end

  def self.try_rare_cases(useragent, result)
    return true if Woothee::Misc.challenge_smartphone_patterns(useragent, result)

    return true if Woothee::Browser.challenge_sleipnir(useragent, result)

    return true if Woothee::Misc.challenge_http_library(useragent, result)

    return true if Woothee::Misc.challenge_maybe_rss_reader(useragent, result)

    return true if Woothee::Crawler.challenge_maybe_crawler(useragent, result)

    false
  end

  FILLED = {
    Woothee::ATTRIBUTE_NAME => Woothee::VALUE_UNKNOWN,
    Woothee::ATTRIBUTE_CATEGORY => Woothee::VALUE_UNKNOWN,
    Woothee::ATTRIBUTE_OS => Woothee::VALUE_UNKNOWN,
    Woothee::ATTRIBUTE_VERSION => Woothee::VALUE_UNKNOWN,
    Woothee::ATTRIBUTE_VENDOR => Woothee::VALUE_UNKNOWN,
  }.freeze
  def self.fill_result(result)
    FILLED.merge(result)
  end
end

File: js/lib/woothee.js
/* HEADER BEGIN */
var dataset = require('./dataset'),
    browser = require('./browser'),
    os = require('./os'),
    mobilephone = require('./mobilephone'),
    crawler = require('./crawler'),
    appliance = require('./appliance'),
    misc = require('./misc');
/* HEADER END */

var parse = exports.parse = function(userAgent) {
  return fillResult(execParse(userAgent));
};

var isCrawler = exports.isCrawler = function(userAgent) {
  return false;
};

function execParse(userAgent) {
  var result = {};

  if (userAgent.length < 1 || userAgent === '-')
    return result;

  if (tryCrawler(userAgent, result))
    return result;

  if (tryBrowser(userAgent, result)) {
    if (tryOS(userAgent, result))
      return result;
    else
      return result;
  }

  if (tryMobilePhone(userAgent, result))
    return result;

  if (tryAppliance(userAgent, result))
    return result;

  if (tryMisc(userAgent, result))
    return result;

  // browser unknown, check os only
  if (tryOS(userAgent, result))
    return result;

  if (tryRareCases(userAgent, result))
    return result;

  return result;
};

function tryCrawler(userAgent, result) {
  if (crawler.challengeGoogle(userAgent, result))
    return true;

  if (crawler.challengeCrawlers(userAgent, result))
    return true;

  return false;
}

function tryBrowser(userAgent, result) {
  if (browser.challengeMSIE(userAgent, result))
    return true;

  if (browser.challengeSafariChrome(userAgent, result))
    return true;

  if (browser.challengeFirefox(userAgent, result))
    return true;

  if (browser.challengeOpera(userAgent, result))
    return true;

  return false;
}

function tryOS(userAgent, result) {
  if (os.challengeWindows(userAgent, result))
    return true;

  // OSX PC and iOS devices (strict check)
  if (os.challengeOSX(userAgent, result))
    return true;

  // Linux PC and Android
  if (os.challengeLinux(userAgent, result))
    return true;

  // all useragents matches /(iPhone|iPad|iPod|Android|BlackBerry)/
  if (os.challengeSmartPhone(userAgent, result))
    return true;

  // mobile phones like KDDI-.*
  if (os.challengeMobilePhone(userAgent, result))
    return true;

  // Nintendo DSi/Wii with Opera
  if (os.challengeAppliance(userAgent, result))
    return true;

  if (os.challengeMisc(userAgent, result))
    return true;

  return false;
}

function tryMobilePhone(userAgent, result) {
  if (mobilephone.challengeDocomo(userAgent, result))
    return true;

  if (mobilephone.challengeAu(userAgent, result))
    return true;

  if (mobilephone.challengeSoftbank(userAgent, result))
    return true;

  if (mobilephone.challengeWillcom(userAgent, result))
    return true;

  if (mobilephone.challengeMisc(userAgent, result))
    return true;

  return false;
}

function tryAppliance(userAgent, result) {
  if (appliance.challengePlaystation(userAgent, result))
    return true;

  if (appliance.challengeNintendo(userAgent, result))
    return true;

  if (appliance.challengeDigitalTV(userAgent, result))
    return true;

  return false;
}

function tryMisc(userAgent, result) {
  if (misc.challengeDesktopTools(userAgent, result))
    return true;

  return false;
}

function tryRareCases(userAgent, result) {
  if (misc.challengeSmartPhonePatterns(userAgent, result))
    return true;

  if (browser.challengeSleipnir(userAgent, result))
    return true;

  if (misc.challengeHTTPLibrary(userAgent, result))
    return true;

  if (misc.challengeMaybeRSSReader(userAgent, result))
    return true;

  if (crawler.challengeMaybeCrawler(userAgent, result))
    return true;

  return false;
}

var filled = {};
filled[dataset.ATTRIBUTE_NAME] = dataset.VALUE_UNKNOWN;
filled[dataset.ATTRIBUTE_CATEGORY] = dataset.VALUE_UNKNOWN;
filled[dataset.ATTRIBUTE_OS] = dataset.VALUE_UNKNOWN;
filled[dataset.ATTRIBUTE_VERSION] = dataset.VALUE_UNKNOWN;
filled[dataset.ATTRIBUTE_VENDOR] = dataset.VALUE_UNKNOWN;

function fillResult(result) {
  if (! result[dataset.ATTRIBUTE_NAME])
    result[dataset.ATTRIBUTE_NAME] = dataset.VALUE_UNKNOWN;
  if (! result[dataset.ATTRIBUTE_CATEGORY])
    result[dataset.ATTRIBUTE_CATEGORY] = dataset.VALUE_UNKNOWN;
  if (! result[dataset.ATTRIBUTE_OS])
    result[dataset.ATTRIBUTE_OS] = dataset.VALUE_UNKNOWN;
  if (! result[dataset.ATTRIBUTE_VERSION])
    result[dataset.ATTRIBUTE_VERSION] = dataset.VALUE_UNKNOWN;
  if (! result[dataset.ATTRIBUTE_VENDOR])
    result[dataset.ATTRIBUTE_VENDOR] = dataset.VALUE_UNKNOWN;
  return result;
};